enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,411 @@
+- If the source CVO image is a manifestlist, the generated release payload would be a manifestlist(containing a manifest for each architecture
+possessed by the CVO manifestlist).
+
+In either case, oc adm release new would permit non-CVO component images to be manifest or manifestlists and pass them through directly to the
+resultant release manifest(s). If a manifestlist release payload is generated, each architecture specific release payload manifest will reference
+the same pullspecs provided in the input imagestream. Since the internal Openshift registry does not support manifestlists, the release controller
+would use quay.io.
+
+In case of the `oc extract` command, it would:
+- by default, extract the tools that will run on the platform on which the oc command is run
+- with `--command-os=*`, extract the tools for all architectures provided by the payload
+
+The above behavior is followed for homogeneous architecture payloads today and it is expected that oc would support it for heterogeneous payloads too.
+The only difference for heterogeneous payloads though is that, the extract command would return a set of tools for all supported payload architectures
+when the `--command-os=*` flag is used. The table below describes the combination of os/architectures for which the oc client and the installer binary
+would be extracted.
+
+| Client platform |	Image arch | Image | Path | Tarball |
+| :--- | :--- | :--- | :--- | :--- |
+| linux/amd64 |	amd64 |	cli | /usr/bin/oc |	openshift-client-linux-amd64-%s.tar.gz |
+| linux/arm64	| arm64	| cli	| /usr/bin/oc | openshift-client-linux-arm64-%s.tar.gz |
+| linux/ppc64le | ppc64le | cli | /usr/bin/oc | openshift-client-linux-ppc64le-%s.tar.gz |
+| linux/s390x | s390x | cli | /usr/bin/oc | openshift-client-linux-s390x-%s.tar.gz |
+| darwin/amd64 | amd64* | cli-artifacts | /usr/share/openshift/mac/oc | openshift-client-mac-%s.tar.gz |
+| darwin/arm64 | arm64* |	cli-artifacts |	/usr/share/openshift/mac_arm64/oc | openshift-client-mac-arm64-%s.tar.gz |
+| windows/amd64 |	amd64* | cli-artifacts | /usr/share/openshift/windows/oc.exe | openshift-client-windows-%s.tar.gz |
+
+| Install platform | Image arch | Image |	Path | Tarball |
+| :--- | :--- | :--- | :--- | :--- |
+| linux/amd64 |	amd64 | installer |	/usr/bin/openshift-install | openshift-install-linux-amd64-%s.tar.gz |
+| linux/arm64	| arm64	| installer | /usr/bin/openshift-install | openshift-install-linux-arm64-%s.tar.gz |
+| linux/ppc64le |	ppc64le | installer | /usr/bin/openshift-install | openshift-install-linux-ppc64le-%s.tar.gz |
+| linux/s390x	| s390x	| installer	| /usr/bin/openshift-install | openshift-install-linux-s390x-%s.tar.gz |
+| darwin/amd64 | amd64* | installer-artifacts |	/usr/share/openshift/mac/openshift-install | openshift-install-mac-%s.tar.gz |
+| darwin/arm64 | arm64* | installer-artifacts | /usr/share/openshift/mac_arm64/openshift-install | openshift-install-mac-arm64-%s.tar.gz |
```

Comment:
A side effect of having a single installer for all payload architectures means that we can no longer rely on hardcoding the payload architectures at build time (as implemented in https://github.com/openshift/installer/pull/4870 and https://github.com/openshift/installer/pull/5107).  Instead, `compute.architecture` and `controlPlane.architecture` will need to be mandatory with no default, and "openshift-install create install-config" modified to add a question to select an architecture among those supported by the selected platform (e.g. `aws` => `['amd64','arm64']`).

Reply:
hmm..yes true..i need to add this as well..thanks for catching this @yselkowitz . also related is - how would the `oc extract` command work? would it extract oc/installer binaries from each of the manifest images in the manifestlisted cli/installer artifacts image? the extract would also have to embed the heterogeneous release payload pullspec in the installer binaries.